src/vm/hints/squash_dict_utils.rs

```diff
@@ -0,0 +1,2485 @@
+use std::collections::HashMap;
+
+use num_bigint::BigInt;
+use num_traits::{FromPrimitive, ToPrimitive};
+
+use crate::{
+    bigint, bigintusize,
+    serde::deserialize_program::ApTracking,
+    types::{exec_scope::PyValueType, relocatable::MaybeRelocatable},
+    vm::{errors::vm_errors::VirtualMachineError, vm_core::VirtualMachine},
+};
+
+use super::{
+    dict_hint_utils::DICT_ACCESS_SIZE,
+    hint_utils::{
+        get_address_from_var_name, get_int_from_scope, get_list_from_scope, get_range_check_builtin,
+    },
+};
+
+fn get_access_indices(vm: &mut VirtualMachine) -> Option<HashMap<BigInt, Vec<BigInt>>> {
+    let mut access_indices: Option<HashMap<BigInt, Vec<BigInt>>> = None;
+    if let Some(variables) = vm.exec_scopes.get_local_variables() {
+        if let Some(PyValueType::KeyToListMap(py_access_indices)) = variables.get("access_indices")
+        {
+            access_indices = Some(py_access_indices.clone());
+        }
+    }
+    access_indices
+}
+
+/*Implements hint:
+    current_access_indices = sorted(access_indices[key])[::-1]
+    current_access_index = current_access_indices.pop()
+    memory[ids.range_check_ptr] = current_access_index
+*/
+pub fn squash_dict_inner_first_iteration(
+    vm: &mut VirtualMachine,
+    ids: HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    //Check that access_indices and key are in scope
+    let access_indices = get_access_indices(vm)
+        .ok_or_else(|| VirtualMachineError::NoLocalVariable(String::from("access_indices")))?;
+    let key = get_int_from_scope(vm, "key")
+        .ok_or_else(|| VirtualMachineError::NoLocalVariable(String::from("key")))?;
+    //Get addr for ids variables
+    let range_check_ptr_addr =
+        get_address_from_var_name("range_check_ptr", &ids, vm, hint_ap_tracking)?;
+    //Get ids from memory
+    let range_check_ptr = vm
+        .memory
+        .get(&range_check_ptr_addr)
+        .map_err(VirtualMachineError::MemoryError)?
+        .ok_or(VirtualMachineError::MemoryGet(range_check_ptr_addr))?;
+    //Get current_indices from access_indices
+    let mut current_access_indices = access_indices
+        .get(&key)
+        .ok_or_else(|| VirtualMachineError::NoKeyInAccessIndices(key.clone()))?
+        .clone();
+    current_access_indices.sort();
+    current_access_indices.reverse();
+    //Get current_access_index
+    let first_val = current_access_indices
+        .pop()
+        .ok_or(VirtualMachineError::EmptyCurrentAccessIndices)?;
```

Comment:
We could just keep a heap in the scope, since we're extracting one item per iteration. Right now this is `O(n^2logn)`, while the heap makes it `O(nlogn)`. The semantics of the hint are still preserved.

Reply:
Or even create a heap locally, at least that's locally linear and ultimately quadratic.